### PR TITLE
Warn when a vulnerable or EOL SDK version is resolved

### LINKED ISSUE #49552: `dotnet` should warn if a vulnerable SDK version is resolved
### Is your feature request related to a problem? Please describe.
As part of most `dotnet` commands, MSBuild resolves the version of the SDK to use. As I understand it, the chain is MSBuild -> `DotNetMSBuildSdkResolver` -> `NETCoreSdkResolver` -> `NETCoreSdkResolverNativeWrapper`. It chooses the most compatible SDK based on: Available SDK versions, MSBuild version compatibility, `global.json` requirements, and preview/release preferences.

I have no easy way of knowing if the resolved SDK version has any applicable vulnerabilities. For an example of what I mean, see [NuGet warnings `NU1901`-`NU1904`][1], which warn if any resolved packages have low-high severity security advisories respectively.

### Describe the solution you'd like
I'd like to see warnings printed to the console if a .NET SDK with a known vulnerability is used. This information is already available as part of the [`releases.json` for each version][2]. For example, for `9.0.6`:

```json
"cve-list": [
  {
    "cve-id": "CVE-2025-30399",
    "cve-url": "https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2025-30399"
  }
]
```

It would also be nice to have a warning printed if an EOL version of the SDK is resolved. This information is available in the [`releases-index.json`][3]:

```json
{
  "channel-version": "7.0",
  "latest-release": "7.0.20",
  "latest-release-date": "2024-05-28",
  "security": true,
  "latest-runtime": "7.0.20",
  "latest-sdk": "7.0.410",
  "product": ".NET",
  "support-phase": "eol",
  "eol-date": "2024-05-14",
  "release-type": "sts",
  "releases.json": "https://builds.dotnet.microsoft.com/dotnet/release-metadata/7.0/releases.json",
  "supported-os.json": "https://builds.dotnet.microsoft.com/dotnet/release-metadata/7.0/supported-os.json"
},
```

### Additional context

[1]: https://learn.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu1901-nu1904
[2]: https://builds.dotnet.microsoft.com/dotnet/release-metadata/9.0/releases.json
[3]: https://builds.dotnet.microsoft.com/dotnet/release-metadata/releases-index.json

### PR DESCRIPTION
**Pull request by JamieMagee**
Closes #49552

NuGet warns about vulnerable packages (NU1901-NU1904) but nothing warns about the SDK itself. This adds opt-in build warnings when the resolved SDK has known CVEs or is end of life.

Set `<CheckSdkVulnerabilities>true</CheckSdkVulnerabilities>` in your project or Directory.Build.props. Off by default. Two warning codes: NETSDK1237 for vulnerabilities, NETSDK1238 for EOL. Suppressible with `<NoWarn>` if you want one but not the other.

During restore, the CLI fetches release metadata in the background (same releases-index.json that `dotnet sdk check` uses) and caches it to `~/.dotnet/sdk-vulnerability-cache/`. The MSBuild task reads from that cache during build. It never hits the network itself. No cache file means no warnings, so air-gapped builds work fine.

Cache refresh runs alongside the workload manifest updater in `RestoringCommand`, 24h sentinel. Skipped in source-build. `DOTNET_SDK_VULNERABILITY_CHECK_DISABLE` and `DOTNET_SDK_VULNERABILITY_CHECK_INTERVAL_HOURS` are there if you need them.

```
warning NETSDK1237: The current .NET SDK (9.0.100) has known vulnerabilities (CVE-2025-12345, CVE-2025-99999). Update to version 9.0.102: https://dotnet.microsoft.com/download
```

```
warning NETSDK1238: The current .NET SDK (7.0.410) is end of life as of 2024-05-14. It will receive no further security updates: https://dotnet.microsoft.com/download
```

@baronfel would appreciate your input.

### CODE REVIEW

**Comment by marcpopMSFT** [comment]
Waiting for @baronfel's comment. We were discussing adding an experience for this once we had a way to update the SDK rather than going to the download page. We were hesitant as we didn't want to have to download releases.json on every CLI action so have not fully designed this feature yet.

**Comment by marcpopMSFT** [comment]
Summarizing offline conversation:
As noted, we were thinking of heading in this direction once we had a way of helping customers update. Some key feedback on the design.

- We think it should only be during build commands, not all commands (still not 100% settled on this)
- As such, it should probably be suppressible through a build property and not just an environment variable. That way you can disable with a checked in change to the diretory.build.props rather than a build script change
- It has to be disableable or be able to override the message in source build builds as the current message doesn't apply to those installs (it doesn't apply to installs by VS or VSCode either fwiw). CC @mthalman @omajid for their feedback on how this type of message could work for source build versions.
- It might be worth exploring if we can tell how the SDK was installed so that the message is customized (or silent) in installs it doesn't make sense for.
- There is some discussion on whether we want it every time but nuget audit does so maybe leave it like that.

Overall, we like the direction of not hitting the network and using the cache instead and it makes sense to warn the user when their out of date. VS and VSCode will have more experiences like this in the future but they're going to have ways of upgrading the user for them which greatly improves the user experience.

**Comment by JamieMagee** [comment]
@marcpopMSFT Thanks for the feedback. I think the right move is to get warnings out of both the CLI startup and the SDK resolver, and put them in a .targets file instead. That one change actually addresses most of your points at once.

Right now warnings fire from `Program.ProcessArgs` (every command) and from the resolver (during build). If I move them into an MSBuild task with dedicated warning codes (`NETSDK2001` for vulnerabilities, `NETSDK2002` for EOL), they'll only show up during build, they'll only show up once, and users can suppress them with `<NoWarn>` or a `<SuppressSdkVulnerabilityWarnings>` property in Directory.Build.props. Same pattern as `SuppressNETCoreSdkPreviewMessage`. The resolver can't do property-based suppression because it runs before project evaluation, which is why I want to move out of it entirely.

The cache refresh moves to `RestoringCommand`, same place as `WorkloadManifestUpdater.BackgroundUpdateAdvertisingManifestsAsync`. So `dotnet --help` and `dotnet tool list` won't touch it.

For source-build, I'll gate the cache refresh with `#if !DOT_NET_BUILD_FROM_SOURCE` so those SDKs skip network calls. On the .targets side, I can set a `_SdkIsSourceBuild` property at build time (in `GenerateBundledVersions.targets`) and either suppress the warning or change the message to say "update via your package manager" instead of linking to dotnet.microsoft.com/download.

For detecting other install sources (VS-managed, install script, etc.), there's no runtime mechanism for that today and adding one means touching MSI, PKG, DEB, RPM, install scripts, and VS bundling. I'd rather defer that to a follow-up. The `DOTNET_SDK_VULNERABILITY_CHECK_DISABLE` env var stays as the escape hatch, and I'll check it in the .targets Condition too so CI and air-gapped builds can opt out.

**Comment by marcpopMSFT** [comment]
More offline conversation:
We discussed this further and are ok taking this as an MSBuild flag but opt-in. While that does limit the scope, it limits the risk as well.

Your suggestions about excluding from source build make sense though with this opt-in, becomes less critical.

Another concern we had would be release day. We don't want there to be a timing problem that ends up breaking people's CI and local builds because the releases.json has updated but their SDK hasn't updated. For example, our own repos usually do a deliberate central rollout with new SDK versions which isn't immediate so this would flag during that validation window.

I think we're still ok having it as an opt-in and we do plan on plugging dotnetup into this experience in the future.

**Comment by JamieMagee** [comment]
@marcpopMSFT Updated the PR.
 
It's opt-in now. `<CheckSdkVulnerabilities>true</CheckSdkVulnerabilities>` in a project or Directory.Build.props enables it, off by default. Warnings use NETSDK1237 (CVEs) and NETSDK1238 (EOL), so `<NoWarn>` works if you want to suppress one but not the other.
 
Warnings moved out of CLI startup and the SDK resolver entirely. They come from a .targets file that runs an MSBuild task during build. `dotnet --help`, `dotnet tool list`, etc. won't trigger anything.
 
Cache refresh moved to `RestoringCommand`, same spot as the workload manifest updater. Background, fire-and-forget, skipped in source-build.
 
On the release-day timing: since it's opt-in the blast radius is already small, but worth noting that CVE warnings only fire when there's a newer release with security fixes in the same channel. So the window only matters if someone opts in, their SDK version is in releases.json, and a newer patch exists they haven't rolled to yet. Repos doing a deliberate rollout could suppress the code in Directory.Build.props during that window.

File: src/Tasks/Microsoft.NET.Build.Tasks/CheckSdkVulnerabilities.cs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using Microsoft.Build.Framework;
+using Microsoft.DotNet.Configurer;
+
+namespace Microsoft.NET.Build.Tasks;
+
+/// <summary>
+/// MSBuild task that reads cached SDK vulnerability/EOL data and emits
+/// NETSDK1236 (vulnerabilities) and NETSDK1237 (EOL) warnings.
```
[the remark this answers is not in the document]
Comment (JamieMagee) [reply src/Tasks/Microsoft.NET.Build.Tasks/CheckSdkVulnerabilities.cs]: Fixed. Leftover from before the rebase bumped the codes.

File: src/Cli/dotnet/SdkVulnerability/SdkReleaseMetadataCache.cs [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        }
+
+        if (File.Exists(sentinelPath))
+        {
+            var lastAccessTime = File.GetLastAccessTime(sentinelPath);
+            if (lastAccessTime.AddHours(updateIntervalHours) > DateTime.Now)
+            {
+                return false;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (JamieMagee) [reply src/Cli/dotnet/SdkVulnerability/SdkReleaseMetadataCache.cs:67]: Switched to `GetLastWriteTimeUtc` / `DateTime.UtcNow`.

File: src/Cli/dotnet/SdkVulnerability/SdkReleaseMetadataCache.cs [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        {
+            var sentinelPath = Path.Combine(_cacheDirectory, SentinelFileName);
+            Directory.CreateDirectory(_cacheDirectory);
+            if (File.Exists(sentinelPath))
+            {
+                File.SetLastAccessTime(sentinelPath, DateTime.Now);
+            }
+            else
+            {
+                File.Create(sentinelPath).Close();
```
[the remark this answers is not in the document]
Comment (JamieMagee) [reply src/Cli/dotnet/SdkVulnerability/SdkReleaseMetadataCache.cs]: Fixed alongside the read side. Both use `LastWriteTimeUtc` + `UtcNow` now.

File: test/dotnet.Tests/SdkVulnerabilityTests/GivenSdkVulnerabilityCacheReader.cs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    [Fact]
+    public void ReturnsNullWhenNoCacheFileExists()
+    {
+        var result = SdkVulnerabilityCacheReader.ReadSummary(
+            "99.99.999",
+            _ => null,
+            "/nonexistent");
+
```
[the remark this answers is not in the document]
Comment (JamieMagee) [reply test/dotnet.Tests/SdkVulnerabilityTests/GivenSdkVulnerabilityCacheReader.cs:35]: Switched to `Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString("N"))`.

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets [hunk 5]
```diff
@@ -87,6 +87,9 @@ Copyright (c) .NET Foundation. All rights reserved.
   <!-- Checks for deprecated Aspire workload usage -->
   <Import Project="$(MSBuildThisFileDirectory)Microsoft.NET.AspireWorkloadDeprecation.targets" />
 
+  <!-- Checks for SDK vulnerabilities and EOL status (opt-in) -->
+  <Import Project="$(MSBuildThisFileDirectory)Microsoft.NET.SdkVulnerabilityCheck.targets" />
+
```
[the remark this answers is not in the document]
Comment (JamieMagee) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets:92]: Updated the description to match. CLI startup and resolver integration were removed in the rework.

File: src/Common/SdkVulnerabilityCacheReader.cs [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+            return null;
+        }
+
+        if (string.IsNullOrEmpty(userProfileDir))
+        {
+            return null;
+        }
+
+        string summaryPath = Path.Combine(userProfileDir, CacheDirectoryName, $"{SummaryFilePrefix}{sdkVersion}.json");
+        if (!File.Exists(summaryPath))
```
[the remark this answers is not in the document]
Comment (JamieMagee) [reply src/Common/SdkVulnerabilityCacheReader.cs:49]: Added `Path.GetInvalidFileNameChars()` validation in both `SdkVulnerabilityCacheReader.ReadSummary` and `CheckSdkVulnerabilities.ExecuteCore` before constructing the path.

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.SdkVulnerabilityCheck.targets [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+  <PropertyGroup>
+    <!-- Opt-in: set to true to check for SDK vulnerabilities and EOL status during build.
+         Suppressible via NoWarn with NETSDK1237 (vulnerabilities) or NETSDK1238 (EOL). -->
+    <CheckSdkVulnerabilities Condition="'$(CheckSdkVulnerabilities)' == ''">false</CheckSdkVulnerabilities>
+  </PropertyGroup>
+
+  <Target Name="_CheckForSdkVulnerabilities" AfterTargets="_CheckForUnsupportedNETCoreVersion"
+          Condition="'$(CheckSdkVulnerabilities)' == 'true'">
+    <CheckSdkVulnerabilities SdkVersion="$(NETCoreSdkVersion)" />
+  </Target>
```
[the remark this answers is not in the document]
Comment (JamieMagee) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.SdkVulnerabilityCheck.targets:24]: Added `<UsingTask TaskName="CheckSdkVulnerabilities" AssemblyFile="$(MicrosoftNETBuildTasksAssembly)" />`.

File: test/dotnet.Tests/SdkVulnerabilityTests/GivenSdkReleaseMetadataCache.cs [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+    [Fact]
+    public void IsDueForUpdateWhenSentinelIsOld()
+    {
+        string cacheDir = GetTempCacheDir();
+        Directory.CreateDirectory(cacheDir);
+        string sentinelPath = Path.Combine(cacheDir, ".sentinel");
+        File.Create(sentinelPath).Close();
+        File.SetLastAccessTime(sentinelPath, DateTime.Now.AddHours(-25));
+
```
[the remark this answers is not in the document]
Comment (JamieMagee) [reply test/dotnet.Tests/SdkVulnerabilityTests/GivenSdkReleaseMetadataCache.cs:77]: Updated all three test methods to use `SetLastWriteTimeUtc + DateTime.UtcNow`.

File: src/Tasks/Microsoft.NET.Build.Tasks/CheckSdkVulnerabilities.cs [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+            return;
+        }
+
+        if (summary.IsEol && summary.EolDate.HasValue)
+        {
+            string message = string.Format(
+                Strings.SdkVersionIsEol,
+                SdkVersion,
+                summary.EolDate.Value.ToString("yyyy-MM-dd"));
+            Log.LogWarning(message);
```
[the remark this answers is not in the document]
Comment (JamieMagee) [reply src/Tasks/Microsoft.NET.Build.Tasks/CheckSdkVulnerabilities.cs]: Fixed. Passing the format string and args directly to `Log.LogWarning()` now.

File: src/Cli/dotnet/SdkVulnerability/SdkReleaseMetadataCache.cs [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+
+        SdkVulnerabilityInfo? info = SdkVulnerabilityChecker.Check(sdkVersion, productCollection, _ => releases);
+        if (info is null)
+        {
+            return;
+        }
+
+        WriteSummary(sdkVersion, info);
+        TouchSentinel();
+    }
```
[the remark this answers is not in the document]
Comment (JamieMagee) [reply src/Cli/dotnet/SdkVulnerability/SdkReleaseMetadataCache.cs:145]: Moved `TouchSentinel()` inside the try block in `WriteSummary`, after the `File.Move` succeeds. Failed writes won't mark the cache as fresh anymore.

File: src/Cli/dotnet/SdkVulnerability/SdkReleaseMetadataCache.cs [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+    /// Reads a previously cached vulnerability summary for the given SDK version.
+    /// Returns null if no cached data is available.
+    /// </summary>
+    public SdkVulnerabilityInfo? ReadCachedSummary(string sdkVersion)
+    {
+        var summaryPath = GetSummaryPath(sdkVersion);
+        if (!File.Exists(summaryPath))
+        {
+            return null;
+        }
```
[the remark this answers is not in the document]
Comment (JamieMagee) [reply src/Cli/dotnet/SdkVulnerability/SdkReleaseMetadataCache.cs:87]: Added the same `Path.GetInvalidFileNameChars()` check that the cache reader and MSBuild task already have.

File: src/Tasks/Common/Resources/xlf/Strings.de.xlf [hunk 12]
```diff
@@ -906,6 +906,21 @@ The following are names of parameters or literal values and should not be transl
         <target state="translated">NETSDK1185: Das Runtimepaket für FrameworkReference „{0}“ war nicht verfügbar. Dies kann daran liegen, dass DisableTransitiveFrameworkReferenceDownloads auf TRUE festgelegt wurde.</target>
         <note>{StrBegins="NETSDK1185: "}</note>
       </trans-unit>
+      <trans-unit id="SdkVersionHasVulnerabilities">
+        <source>NETSDK1237: The current .NET SDK ({0}) has known vulnerabilities ({1}). Update to version {2}: https://dotnet.microsoft.com/download</source>
+        <target state="new">NETSDK1237: The current .NET SDK ({0}) has known vulnerabilities ({1}). Update to version {2}: https://dotnet.microsoft.com/download</target>
+        <note>{StrBegins="NETSDK1237: "}{Locked="{0}"}{Locked="{1}"}{Locked="{2}"}</note>
+      </trans-unit>
```
[the remark this answers is not in the document]
Comment (JamieMagee) [reply src/Tasks/Common/Resources/xlf/Strings.de.xlf:923]: These were generated by `msbuild /t:UpdateXlf`, not hand-edited.

File: src/Tasks/Microsoft.NET.Build.Tasks/CheckSdkVulnerabilities.cs [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+        }
+
+        // Validate version to prevent path traversal via malformed version strings
+        if (SdkVersion.IndexOfAny(Path.GetInvalidFileNameChars()) >= 0)
+        {
+            return;
+        }
+
+        string summaryPath = Path.Combine(userProfileDir, CacheDirectoryName, $"{SummaryFilePrefix}{SdkVersion}.json");
+        if (!File.Exists(summaryPath))
```
[the remark this answers is not in the document]
Comment (JamieMagee) [reply src/Tasks/Microsoft.NET.Build.Tasks/CheckSdkVulnerabilities.cs:46]: Added. Task now returns early if the env var is set, same as the cache refresher and cache reader.

File: test/dotnet.Tests/SdkVulnerabilityTests/GivenSdkReleaseMetadataCache.cs [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+        var cache1h = new SdkReleaseMetadataCache(
+            cacheDir,
+            name => name == EnvironmentVariableNames.SDK_VULNERABILITY_CHECK_INTERVAL_HOURS ? "1" : null);
+        cache1h.IsDueForUpdate().Should().BeTrue();
+    }
+
+    private string GetTempCacheDir()
+    {
+        return Path.Combine(Path.GetTempPath(), "sdk-vuln-test-" + Guid.NewGuid().ToString("N")[..8]);
+    }
```
[the remark this answers is not in the document]
Comment (JamieMagee) [reply test/dotnet.Tests/SdkVulnerabilityTests/GivenSdkReleaseMetadataCache.cs:158]: Class implements `IDisposable` now. `GetTempCacheDir()` tracks every dir it creates, `Dispose()` deletes them.

File: src/Cli/dotnet/SdkVulnerability/SdkVulnerabilityNotifier.cs [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+        {
+            sdkVersion ??= Product.Version;
+
+            var cache = new SdkReleaseMetadataCache();
+
+            if (cache.IsDisabled() || !cache.IsDueForUpdate())
+            {
+                return;
+            }
+
```
[the remark this answers is not in the document]
Comment (JamieMagee) [reply src/Cli/dotnet/SdkVulnerability/SdkVulnerabilityNotifier.cs]: The notifier now also checks if there's a cached summary for the current SDK version. If the sentinel is fresh but there's no `sdk-status-{version}.json`, it triggers a refresh anyway.

File: src/Cli/dotnet/SdkVulnerability/SdkReleaseMetadataCache.cs [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+    /// Returns true if the sentinel file indicates the cache is due for a refresh.
+    /// </summary>
+    public bool IsDueForUpdate()
+    {
+        var sentinelPath = Path.Combine(_cacheDirectory, SentinelFileName);
+
+        if (!int.TryParse(_getEnvironmentVariable(EnvironmentVariableNames.SDK_VULNERABILITY_CHECK_INTERVAL_HOURS), out int updateIntervalHours))
+        {
+            updateIntervalHours = DefaultUpdateIntervalHours;
+        }
```
[the remark this answers is not in the document]
Comment (JamieMagee) [reply src/Cli/dotnet/SdkVulnerability/SdkReleaseMetadataCache.cs]: Clamped to minimum 1. Values below that fall back to the 24h default.

File: src/Cli/dotnet/SdkVulnerability/SdkReleaseMetadataCache.cs [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+        {
+            releases = [];
+        }
+        cancellationToken.ThrowIfCancellationRequested();
+
+        SdkVulnerabilityInfo? info = SdkVulnerabilityChecker.Check(sdkVersion, productCollection, _ => releases);
+        if (info is null)
+        {
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (JamieMagee) [reply src/Cli/dotnet/SdkVulnerability/SdkReleaseMetadataCache.cs:142]: All the early-return paths (unparseable version, unknown channel, null from checker) now call `TouchSentinel()` so we don't hammer the network on every restore for an SDK that simply isn't in releases.json.

File: src/Tasks/Microsoft.NET.Build.Tasks/CheckSdkVulnerabilities.cs [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+{
+    private const string CacheDirectoryName = "sdk-vulnerability-cache";
+    private const string SummaryFilePrefix = "sdk-status-";
+
+    [Required]
+    public string SdkVersion { get; set; } = string.Empty;
+
+    protected override void ExecuteCore()
+    {
+        if (bool.TryParse(Environment.GetEnvironmentVariable("DOTNET_SDK_VULNERABILITY_CHECK_DISABLE"), out bool disabled) && disabled)
```
[the remark this answers is not in the document]
Comment (JamieMagee) [reply src/Tasks/Microsoft.NET.Build.Tasks/CheckSdkVulnerabilities.cs]: Using `Microsoft.DotNet.Cli.EnvironmentVariableNames.SDK_VULNERABILITY_CHECK_DISABLE` now.

**Comment by JoeRobich** [comment]
Thinking of the offline install scenario, could the SDK installer ship a copy of the vulnerability data and update the local cache if it is newer?

**Comment by JamieMagee** [comment]
@JoeRobich Interesting idea, but I think there's a timing problem. The 11.0.100 installer is built when 11.0.100 is the latest release, so the bundled `releases.json` would show it as current with zero CVEs. Vulnerabilities get disclosed later in 11.0.101 or 11.0.102, and by then the 11.0.100 installer has already shipped. Installers aren't rebuilt after the fact, so the bundled data would be out of date in exactly the scenarios you'd want it for.

The current design already handles offline. No cache file means no warnings, no errors, just a no-op.

**Review by brettfo — APPROVED** [review]

**Comment by baronfel** [comment]
Marked as do-not-merge because this hasn't been reviewed by the SDK team yet.

File: src/Cli/dotnet/SdkVulnerability/SdkVulnerabilityChecker.cs [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+internal static class SdkVulnerabilityChecker
+{
+    /// <summary>
+    /// Checks the given SDK version against the product collection and release data.
+    /// </summary>
+    /// <param name="sdkVersionString">The resolved SDK version string (e.g. "9.0.300").</param>
+    /// <param name="productCollection">The product collection from releases-index.json.</param>
+    /// <param name="getProductReleases">A function to get releases for a product (may read from cache).</param>
+    /// <returns>Vulnerability info, or null if the SDK version cannot be matched to release data.</returns>
+    public static SdkVulnerabilityInfo? Check(
```
Comment (joeloff) [inline src/Cli/dotnet/SdkVulnerability/SdkVulnerabilityChecker.cs:21]: Does this check for discontinued feature band, for example, at some point we typically stop shipping 2xx or 3xx and only 1xx and 4xx would be in the latest release. So, even though the product (e.g. 8.0) is not EOL, we may have stopped producing 3xx. Essentially you'd want to check if the SDKs collection in the latest release contains the current feature band, if not, it's vulnerable, but technically only if there are at least one newer release that contains security fixes. Depends on how strict you want to be. The runtime of a newer release may contain CVEs, but those may not necessarily impact the SDK.

**Comment by joeloff** [comment]
Do the checks handle newer SDKs that are not published? There are a couple of edge cases.

1. New major release of .NET being used before the first preview is shipped - early releases usually have an alpha designation.
2. Daily build of an unreleased SDK, e.g. new feature band that's in preview.
3. Official feature band that's not part of the releases.json. We had a special case with .NET 5 and Visual Studio 16. .NET 5 was going out of support, .NET Core 3.1 was still in support. So we ended up shipping 5.0.4xx SDKs that ran on .NET 6. These only shipped with VS, and allowed users to continue targeting .NET Core 3.1 until newer versions of VS shipped .NET 6.0 SDK.

For early releases, I think it's fine to consider it in support, e.g. if the Major.Minor is newer than the latest product version, e.g. you installed .NET 12.0, but the last official product we shipped was 11.0. Same for SDKs, if the SDK version is newer than any known SDK and the product is not EOL, we should consider the SDK to be valid.

**Comment by joeloff** [comment]
> * It might be worth exploring if we can tell how the SDK was installed so that the message is customized (or silent) in installs it doesn't make sense for.

It's doable, we have code that can do this on Windows, and do it extremely accurately, but it's expensive. If you want that to have value, you have to be accurate.

File: src/Cli/dotnet/SdkVulnerability/SdkVulnerabilityChecker.cs (hunk 19 above)
Comment (JamieMagee) [reply src/Cli/dotnet/SdkVulnerability/SdkVulnerabilityChecker.cs:21]: Right now if someone's on a discontinued band, the checker still pulls in CVEs from newer releases (since those releases are newer than theirs) but `FindLatestSdkInFeatureBand` returns null because it only looks within the same band. So they get a "has known vulnerabilities, no upgrade available" warning pointing at the download page, which is misleading because there *is* an upgrade, they just need to switch bands.

I'll add a discontinued-band path: if there's no newer SDK in the user's band but `product.LatestSdkVersion` is greater than theirs, recommend that instead and use a new warning string saying the band is no longer serviced. Should just be a bool + string on `SdkVulnerabilityInfo`.

On runtime vs SDK CVE attribution, I looked and don't see a clean way to do it. The `cves` collection is per-release, not per-component, and there's no flag for runtime-only fixes. The SDK ships the runtime anyway, so a runtime CVE in your SDK's bundled runtime is genuinely a CVE in your install. I'd rather over-warn than try to parse CVE descriptions to guess scope. Let me know if you have a signal I'm missing though.

On strictness I'm leaning towards warn whenever a same-channel newer release with CVEs exists. Stricter than what you described but matches what I'd expect from a "vulnerable SDK" warning. Happy to dial it back if that's too aggressive.

**Comment by JamieMagee** [comment]
> Do the checks handle newer SDKs that are not published? There are a couple of edge cases.

All three are already handled. The `currentReleaseIndex < 0 -> return null` early exit covers 2 and 3, and case 1 falls out from the channel lookup returning null.

> 1. New major release of .NET being used before the first preview is shipped - early releases usually have an alpha designation.

Case 1, 12.0 alpha when latest released is 11.0: channel `"12.0"` isn't in releases-index.json, checker returns null. Covered by `UnknownSdkVersionReturnsNull`.

> 2. Daily build of an unreleased SDK, e.g. new feature band that's in preview.

Case 2, daily build or new in-preview band on a known channel: channel found, SDK not in any release on it, `currentReleaseIndex < 0`, returns null. Covered by `KnownChannelButUnknownSdkVersionReturnsNull`.

> 3. Official feature band that's not part of the releases.json. We had a special case with .NET 5 and Visual Studio 16. .NET 5 was going out of support, .NET Core 3.1 was still in support. So we ended up shipping 5.0.4xx SDKs that ran on .NET 6. These only shipped with VS, and allowed users to continue targeting .NET Core 3.1 until newer versions of VS shipped .NET 6.0 SDK.

Case 3 is the same path as case 2. Channel can be EOL, but if the SDK isn't in releases.json we return null and emit nothing. For the 5.0.4xx scenario that seems right — those SDKs were deliberately shipped to let people target EOL frameworks, so warning about EOL and pointing at the download page would be actively wrong.

I'll add a regression test for case 3 (EOL channel + unknown SDK -> null) so the behaviour doesn't drift.

File: src/Cli/dotnet/SdkVulnerability/SdkVulnerabilityChecker.cs (hunk 19 above)
Comment (joeloff) [reply src/Cli/dotnet/SdkVulnerability/SdkVulnerabilityChecker.cs:21]: No, that sounds good to me. Starting off more aggressive feels like the correct approach given that we're dealing with vulnerabilities.

**Review by joeloff — APPROVED** [review]

**Comment by joeloff** [comment]
@baronfel do we have a tracking issue for docs?

**Comment by JamieMagee** [comment]
@joeloff No tracking issue, but I opened dotnet/docs#53870 with the three error pages and supporting cross-references.

**Comment by serverdev42** [comment]
Can this be back-ported? Docs doesn't mention which version of the SDK these checks are available in so I opened https://github.com/dotnet/docs/issues/55442 for that. Very important that this information is available in your docs pages.
